### PR TITLE
Addresses MultiSpeed AirToAirHP Total Electricity Rate is Double Counting the Supplemental Heater Electricity Rate

### LINKED ISSUE #9840: MultiSpeed AirToAirHP Total Electricity Rate is Double Counting the Supplemental Heater Electricity Rate
Issue overview
--------------
`AirLoopHVAC:UnitaryHeatPump:AirToAir:MultiSpeed` parent object total electricity rate reported as `Unitary System Electricity Rate` double counts the supplemental heater electricity rate. 

### Details
Some additional details for this issue (if relevant):
 - Platform (Any)
 - Version of EnergyPlus (V22.2)

The offending code is [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/HVACMultiSpeedHeatPump.cc#L473-L477) also shown below:

```
        Real64 locFanElecPower = 0.0;
        locFanElecPower = Fans::GetFanPower(state, MSHeatPump(MSHeatPumpNum).FanNum);
        MSHeatPump(MSHeatPumpNum).ElecPower = locFanElecPower + state.dataHVACGlobal->DXElecCoolingPower + state.dataHVACGlobal->DXElecHeatingPower +
                                              state.dataHVACGlobal->ElecHeatingCoilPower + state.dataHVACGlobal->SuppHeatingCoilPower +
                                              MSHeatPump(MSHeatPumpNum).AuxElecPower;
```
The variable `state.dataHVACGlobal->ElecHeatingCoilPower ` already includes `state.dataHVACGlobal->SuppHeatingCoilPower`. per this code [here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/HeatingCoils.cc#L3037-L3041). 
 
### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by Nigusse**
Pull request overview
---------------------
 - Fixes #9840 
 - (1) Fixes `AirLoopHVAC:UnitaryHeatPump:AirToAir:MultiSpeed` parent object total electricity rate reported as `Unitary System Electricity Rate` double counting the supplemental heater electricity rate. 
 - (2) Corrects overriding electricity rate of main electric heating coil when the main and the supplemental heating coils are electric heating coils in `AirLoopHVAC:UnitaryHeatPump:AirToAir:MultiSpeed` object.
 - (3) Fixes `AirLoopHVAC:UnitarySystem` parent object `Unitary System Electricity Rate` report variable missing multi-stage electric heating coil as a supplemental heating coil.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [x] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [x] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [x] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

**Comment by Nigusse** [comment]
The plot below shows comparing of the sum of the components of the Unitary System electricity use rates and the aggregate report variable `Unitary System Electricity Rates` for `develop` and this `branch`.  The branch plot shows the values match.

SupplementalHtgCoil

File: src/EnergyPlus/HeatingCoils.cc [hunk 1]
```diff
@@ -2067,6 +2065,11 @@ namespace HeatingCoils {
             }
         }
 
+        if (state.dataHeatingCoils->CoilIsSuppHeater) {
+            state.dataHVACGlobal->SuppHeatingCoilPower = heatingCoil.ElecUseLoad;
+        } else {
+            state.dataHVACGlobal->ElecHeatingCoilPower = heatingCoil.ElecUseLoad;
+        }
```
Comment (rraustad) [inline src/EnergyPlus/HeatingCoils.cc]: This is already in ReportHeatingCoil so is not needed in the calc routines. It looks to me as if you only needed to delete the reference to ElecHeatingCoilPower, and other code, as you have done.

    void ReportHeatingCoil(EnergyPlusData &state, int const CoilNum, bool const coilIsSuppHeater)
    {

        if (coilIsSuppHeater) {
            state.dataHVACGlobal->SuppHeatingCoilPower = heatingCoil.ElecUseLoad;
        } else {
            state.dataHVACGlobal->ElecHeatingCoilPower = heatingCoil.ElecUseLoad;
        }

Comment (Nigusse) [reply src/EnergyPlus/HeatingCoils.cc]: That was my initial thought but when I looked at the code Unitary System calls the `CalcMultiStageElectricHeatingCoil` function directly in several places and it won't hit the `ReportHeatingCoil` function.

Comment (rraustad) [reply src/EnergyPlus/HeatingCoils.cc]: Yes, that is to cut down on code execution during iterations. But as a final call, after the PLR is known, the parent should call the sim function for all coils and fan, which will call Report. See if that is happening and if not then there is the problem.

**Comment by Nigusse** [comment]
Another case where the main heating coil electricity rate saved in `state.dataHVACGlobal->ElecHeatingCoilPower` is overridden by the supplemental heating coil electricity rate when the main and the supplemental heating coils are electric. The first plots shows `Unitary System Electricity Rate` does not include the main heating coil electricity rate which is the largest component but still double counting the supplemental electric heating coil electricity rate. The second plot (this branch) shows the `Unitary System Electricity Rate` matches the sum of the components.

SupplementalHtgCoil2

File: src/EnergyPlus/HeatingCoils.cc [hunk 2]
```diff
@@ -1937,7 +1937,6 @@ namespace HeatingCoils {
         // Simulates a simple Electric heating coil with an efficiency
 
         // Using/Aliasing
-        auto &ElecHeatingCoilPower = state.dataHVACGlobal->ElecHeatingCoilPower;
```
Comment (Nigusse) [inline src/EnergyPlus/HeatingCoils.cc:1940]: Removed, redundant code. This code overrides the main electric heating coil electricity rate if the supplemental heating coil is also electric heating coil.

File: src/EnergyPlus/HeatingCoils.cc [hunk 3]
```diff
@@ -2042,7 +2041,6 @@ namespace HeatingCoils {
         }
 
         heatingCoil.HeatingCoilLoad = HeatingCoilLoad;
-        ElecHeatingCoilPower = heatingCoil.ElecUseLoad;
```
Comment (Nigusse) [inline src/EnergyPlus/HeatingCoils.cc:2044]: This variable can be set in a single location. This code overrides the main electric heating coil electricity rate if the supplemental heating coil is also electric heating coil.

File: src/EnergyPlus/HeatingCoils.cc [hunk 4]
```diff
@@ -2267,10 +2266,14 @@ namespace HeatingCoils {
             // some of these are reset in Init, can be removed to speed up code
             heatingCoil.ElecUseLoad = 0.0;
             heatingCoil.HeatingCoilLoad = 0.0;
-            ElecHeatingCoilPower = 0.0;
```
Comment (Nigusse) [inline src/EnergyPlus/HeatingCoils.cc:2270]: This variable can be set in a single location. This code overrides the main electric heating coil electricity rate if the supplemental heating coil is also electric heating coil.

File: src/EnergyPlus/HeatingCoils.cc [hunk 5]
```diff
@@ -2098,7 +2101,6 @@ namespace HeatingCoils {
 
         // Using/Aliasing
         using Curve::CurveValue;
-        auto &ElecHeatingCoilPower = state.dataHVACGlobal->ElecHeatingCoilPower;
```
Comment (Nigusse) [inline src/EnergyPlus/HeatingCoils.cc:2101]: Removed, redundant code.

File: src/EnergyPlus/HeatingCoils.cc [hunk 6]
```diff
@@ -2238,8 +2239,6 @@ namespace HeatingCoils {
 
                 heatingCoil.ElecUseLoad = heatingCoil.HeatingCoilLoad / EffLS;
 
-                ElecHeatingCoilPower = heatingCoil.ElecUseLoad;
-
```
Comment (Nigusse) [inline src/EnergyPlus/HeatingCoils.cc:2242]: This variable can be set in a single location. This code overrides the main electric heating coil electricity rate if the supplemental heating coil is also electric heating coil.

File: src/EnergyPlus/HeatingCoils.cc [hunk 7]
```diff
@@ -2267,10 +2266,14 @@ namespace HeatingCoils {
...
             heatingCoil.HeatingCoilLoad = 0.0;
-            ElecHeatingCoilPower = 0.0;
 
         } // end of on/off if - else
 
+        if (state.dataHeatingCoils->CoilIsSuppHeater) {
+            state.dataHVACGlobal->SuppHeatingCoilPower = heatingCoil.ElecUseLoad;
+        } else {
+            state.dataHVACGlobal->ElecHeatingCoilPower = heatingCoil.ElecUseLoad;
+        }
```
Comment (Nigusse) [inline src/EnergyPlus/HeatingCoils.cc]: These two variables are set in a single location depending the `state.dataHeatingCoils->CoilIsSuppHeater` flag variable and eliminates overriding `state.dataHVACGlobal->ElecHeatingCoilPower`.

**Review by Nigusse — COMMENTED** [review]
Code changes walk-through.

File: src/EnergyPlus/HeatingCoils.cc (hunk 1 above)
Comment (Nigusse) [reply src/EnergyPlus/HeatingCoils.cc]: Stepped through the code and confirmed that the report variables update code in `ReportHeatingCoil` is sufficient. Also ran example files and and the unit test for additional verification. I am going to remove the redundant code.

File: src/EnergyPlus/UnitarySystem.cc [hunk 8]
```diff
@@ -16066,7 +16066,8 @@ namespace UnitarySystems {
         }
 
         if (this->m_SuppCoilExists) {
-            if (this->m_HeatingCoilType_Num == DataHVACGlobals::Coil_HeatingElectric) {
+            if (this->m_HeatingCoilType_Num == DataHVACGlobals::Coil_HeatingElectric ||
+                this->m_HeatingCoilType_Num == DataHVACGlobals::Coil_HeatingElectric_MultiStage) {
                 suppHeatingPower = state.dataHVACGlobal->SuppHeatingCoilPower;
             }
         }
```
Comment (Nigusse) [inline src/EnergyPlus/UnitarySystem.cc:16073]: Multistage supplemental electric heating coil electricity rate was not included in the Unitary System parent total electricity rate. This change fixes that.

**Comment by Nigusse** [comment]
Uploaded additional defect file.

File: src/EnergyPlus/UnitarySystem.cc (hunk 8 above)
Comment (Myoldmopar) [reply src/EnergyPlus/UnitarySystem.cc:16073]: Looks good.

File: tst/EnergyPlus/unit/HVACMultiSpeedHeatPump.unit.cc [hunk 9]
```diff
@@ -1502,4 +1503,734 @@ TEST_F(EnergyPlusFixture, HVACMultiSpeedHeatPump_HeatRecoveryTest)
...
+    QZnReq = zoneSysEnergyDemand.RemainingOutputRequired;
+    msHeatPump.HeatCoolMode = EnergyPlus::HVACMultiSpeedHeatPump::ModeOfOperation::HeatingMode;
+    SimMSHP(*state, MSHeatPumpNum, FirstHVACIteration, AirLoopNum, QSensUnitOut, QZnReq, OnOffAirFlowRatio);
+    // calculate the total electricity rate
+    Real64 result_msHeatPump_ElectricityRate = supplyFan.FanPower + state->dataHVACGlobal->DXElecCoolingPower +
+                                               state->dataHVACGlobal->DXElecHeatingPower + state->dataHVACGlobal->ElecHeatingCoilPower +
+                                               state->dataHVACGlobal->SuppHeatingCoilPower + msHeatPump.AuxElecPower;
+    // test results
+    EXPECT_NEAR(55366.46, msHeatPump.ElecPower, 0.01);
+    EXPECT_NEAR(55366.46, result_msHeatPump_ElectricityRate, 0.01);
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/HVACMultiSpeedHeatPump.unit.cc:2224]: This looks to be the actual test of electricity rate.  I personally wouldn't mind the tolerances being loosened slightly.  The simulation accuracy is not meaningful down to the centiwatt, but this is fine for now.  Definitely an item for another day.

**Review by Myoldmopar — APPROVED** [review]

**Comment by Myoldmopar** [comment]
CI was happy the last time this ran.  Pulling develop in locally everything still builds and runs happily.  I'm going to try to run the defect file and this should be ready to merge.

File: src/EnergyPlus/HeatingCoils.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/HeatingCoils.cc]: Looks like this got resolved. :+1:

**Comment by Myoldmopar** [comment]
@Nigusse I took the SingleFamily* defect file and ran it for design days only with develop and this branch.  In the base case, the unitary system electricity rate is ~23kW on the winter design day.  In the mod case it is way lower, ~4kW.  I was expecting things to be "cut in half" since we were double counting.  Am I missing something?  That's just a huge difference.

**Comment by Nigusse** [comment]
Each of the three defect file illustrate different defect. The double counting of supplemental heating coil electricity rate occurs when you have the main heating coil and the supplemental heating coils are electric type. See the defect file `MultiSpeedACElecFurnace_mod.idf` for such configuration. Develop should show that the supplemental electric heating coil electricity rate is added to the main heating coil electricity rate and and the same value is also reported in the supplemental heating coil. When adding the electricity rates of the main and supplemental heat coils to the parent the double counting occurs.

**Comment by Nigusse** [comment]
@Myoldmopar Note that PR #9841 is already merged into PR ##9839 such that enable me verify the parent total electricity rate is equal to the sum of the components when defrost and crankcase heating electricity rate are included. Again, the issues may vary for the different parent HVAC system types addressed here as well.

**Comment by Nigusse** [comment]
Also, the other issue was when the main heating coil electricity rate is non-zero value (> 0) and the supplemental heating coil is off, then the the main heating coil electricity rate (saved in the global variable) was overridden by the supplemental heating coil rate, which is zero. This global variable (which is zero) is added to the parent but the main heating coil electricity rate is still non-zero. That is when you see a huge difference between the parent and sum of the components. Again this happens when the main and the supplemental heating coils are the electric heating coils.

**Comment by Myoldmopar** [comment]
OK I was reviewing the defrost one and trying to resolve conflicts and such over there.  It would seem I need to get this merged in first since that PR includes this one.  I see that you have investigated and identified the issues and provided numerous defect files with results analysis.  I will get this one merged in now and move on to that PR.  Thanks for responding with the details @Nigusse

**Pull request merged**